File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-fixed-size/sliding-window-maximum/code/java/SlidingWindowMaximum.java
// ============================================================
// Sliding Window Maximum
// ============================================================

import java.util.*;

// ============================================================
// Solution 1: Brute Force
// Time: O(N×K) | Space: O(1)
// ============================================================
class Solution1 {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] result = new int[nums.length - k + 1];
        for (int i = 0; i <= nums.length - k; i++) {
            int max = nums[i];
            for (int j = i + 1; j < i + k; j++) max = Math.max(max, nums[j]);
            result[i] = max;
        }
        return result;
    }
}

// ============================================================
// Solution 2: Monotonic Decreasing Deque (Optimal)
// Time: O(N) | Space: O(K)
// ============================================================
class Solution2 {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int[] result = new int[nums.length - k + 1];
        int idx = 0;

        for (int i = 0; i < nums.length; i++) {
            // Remove expired
            while (!dq.isEmpty() && dq.peekFirst() < i - k + 1) dq.pollFirst();

            // Remove smaller
            while (!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]) dq.pollLast();

            dq.offerLast(i);

            if (i >= k - 1) result[idx++] = nums[dq.peekFirst()];
        }

        return result;
    }
}

File: src/data/courses/dsa/queues/deque/sliding-window-maximum/code/cpp/sliding_window_maximum.cpp
// ============================================================
// Sliding Window Maximum
// ============================================================

#include <vector>
#include <deque>
using namespace std;

// ============================================================
// Solution 1: Brute Force — Scan Each Window
// Time: O(N×K) | Space: O(1)
// ============================================================
class Solution1 {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        for (int i = 0; i <= (int)nums.size() - k; i++) {
            int maxVal = nums[i];
            for (int j = i + 1; j < i + k; j++)
                maxVal = max(maxVal, nums[j]);
            result.push_back(maxVal);
        }
        return result;
    }
};

// ============================================================
// Solution 2: Monotonic Decreasing Deque (Optimal)
// Time: O(N) | Space: O(K)
// ============================================================
class Solution2 {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq; // Stores indices; front = current max
        vector<int> result;

        for (int i = 0; i < nums.size(); i++) {
            // Remove expired indices from front
            while (!dq.empty() && dq.front() < i - k + 1)
                dq.pop_front();

            // Remove smaller elements from back
            // (they'll never be max while nums[i] is in the window)
            while (!dq.empty() && nums[dq.back()] <= nums[i])
                dq.pop_back();

            // Add current index
            dq.push_back(i);

            // Window is full — front is the max
            if (i >= k - 1)
                result.push_back(nums[dq.front()]);
        }

        return result;
    }
};
